dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveVisitor.java

```diff
@@ -0,0 +1,71 @@
+package com.ibm.ws.cdi.impl.weld;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Stack;
+
+import com.ibm.ws.cdi.CDIException;
+import com.ibm.ws.cdi.internal.interfaces.WebSphereBeanDeploymentArchive;
+
+import org.apache.commons.collections4.iterators.FilterIterator;
+import org.apache.commons.collections4.iterators.PeekingIterator;
+
+public class BeanDeploymentArchiveVisitor {
+
+    private final Stack<PeekingIterator<WebSphereBeanDeploymentArchive>> stack = new Stack<PeekingIterator<WebSphereBeanDeploymentArchive>>();
+
+    /**
+     * This method works by visitng each BDA twice. The first time it will get a list of all children BDAs; filtered to remove those that have already been visited
+     * The second time it scans the BDA.
+     *
+     * The second time may be immediately. If there are no children the while loop will read the same entry of the stack. Or it may be when the stack rolls back.
+     *
+     * This use of filtered lists ensures that if we have the following parent child relationships
+     * A -> BD
+     * B -> AD
+     * D -> none
+     *
+     * The stack will be:
+     * 1) A
+     * 2) A, BD
+     * 3) A, BD, D
+     *
+     * At which point D will be scanned.
+     * Then B will be scanned. Then D will not be scanned.
+     * Then finally A will be scanned.
+     */
+    public void visit(WebSphereBeanDeploymentArchive root) throws CDIException {
+        List<WebSphereBeanDeploymentArchive> list = new ArrayList<WebSphereBeanDeploymentArchive>(1);
+        list.add(root);
+        stack.add(PeekingIterator.peekingIterator(list.iterator()));
+
+        while (!stack.isEmpty()) {
+
+            PeekingIterator<WebSphereBeanDeploymentArchive> iterator = stack.peek();
+
+            if (iterator.hasNext()) {
+                WebSphereBeanDeploymentArchive bda = iterator.peek();
+                if (!bda.hasBeenVisited()) {
+                    Iterator<WebSphereBeanDeploymentArchive> children = bda.visit();
+                    addNextStack(children);
+                } else {
+                    if (bda.hasBeenScanned()) {
+                        bda.scan();
+                    }
+                    iterator.next();
```

Comment:
I don't think this part works for two reasons:
1. it should be `if (!bda.hasBeenScanned()) {`
2. the `FilterIterator` will filter out the BDA because it's now been visited

Reply:
Regarding two, I checked the code of PeekingIterator before coding this, and the way it works is that when you call peek it calls the decorated iterator and caches the result. So we still have access to the BDA when we need it.

Reply:
Ah! I see, yes, it would match the filter when you call `peek()` the first time, just not the second time.